File: account/src/main/java/serp/project/account/ui/controller/ModuleController.java
package serp.project.account.ui.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import serp.project.account.core.domain.dto.request.CreateModuleDto;
import serp.project.account.core.domain.dto.request.UpdateModuleDto;
import serp.project.account.core.usecase.ModuleUseCase;
import serp.project.account.kernel.utils.AuthUtils;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/v1/modules")
public class ModuleController {
    private final ModuleUseCase moduleUseCase;

    private final AuthUtils authUtils;

    @PostMapping
    public ResponseEntity<?> createModule(@Valid @RequestBody CreateModuleDto request) {
        var response = moduleUseCase.createModule(request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/{moduleId}")
    public ResponseEntity<?> getModuleById(@PathVariable Long moduleId) {
        var response = moduleUseCase.getModuleById(moduleId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PutMapping("/{moduleId}")
    public ResponseEntity<?> updateModule(@PathVariable Long moduleId, @Valid @RequestBody UpdateModuleDto request) {
        var response = moduleUseCase.updateModule(moduleId, request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping
    public ResponseEntity<?> getAllModules() {
        var response = moduleUseCase.getAllModules();
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/{moduleId}/registration")
    public ResponseEntity<?> userRegisterModule(@PathVariable Long moduleId) {
        Long userId = authUtils.getCurrentUserId().orElse(null);
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);

        if (userId == null || tenantId == null) {
            return ResponseEntity.status(401).body("Unauthorized: Missing user or tenant information");
        }

        var response = moduleUseCase.userRegisterModule(userId, moduleId, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/my-modules")
    public ResponseEntity<?> getMyModules() {
        Long userId = authUtils.getCurrentUserId().orElse(null);
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);

        if (userId == null) {
            return ResponseEntity.status(401).body("Unauthorized: Missing user information");
        }

        if (tenantId == null) {
            return ResponseEntity.status(401).body("Unauthorized: Missing tenant information");
        }

        var response = moduleUseCase.getUserModules(userId, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }
}

File: api_gateway/src/ui/controller/account/module_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	request "github.com/serp/api-gateway/src/core/domain/dto/request/account"
	service "github.com/serp/api-gateway/src/core/service/account"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type ModuleController struct {
	moduleService service.IModuleService
}

func (m *ModuleController) CreateModule(c *gin.Context) {
	var req request.CreateModuleDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := m.moduleService.CreateModule(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *ModuleController) GetModuleById(c *gin.Context) {
	moduleId, ok := utils.ValidateAndParseID(c, "moduleId")
	if !ok {
		return
	}

	res, err := m.moduleService.GetModuleById(c.Request.Context(), moduleId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *ModuleController) GetRolesInModule(c *gin.Context) {
	moduleId, ok := utils.ValidateAndParseID(c, "moduleId")
	if !ok {
		return
	}

	res, err := m.moduleService.GetRolesInModule(c.Request.Context(), moduleId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *ModuleController) UpdateModule(c *gin.Context) {
	moduleId, ok := utils.ValidateAndParseID(c, "moduleId")
	if !ok {
		return
	}

	var req request.UpdateModuleDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := m.moduleService.UpdateModule(c.Request.Context(), moduleId, &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *ModuleController) GetAllModules(c *gin.Context) {
	res, err := m.moduleService.GetAllModules(c.Request.Context())
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *ModuleController) GetMyModules(c *gin.Context) {
	res, err := m.moduleService.GetMyModules(c.Request.Context())
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func NewModuleController(moduleService service.IModuleService) *ModuleController {
	return &ModuleController{
		moduleService: moduleService,
	}
}
